presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ActualProperties.java

```diff
@@ -353,6 +373,20 @@ public static Global partitionedOn(Partitioning nodePartitioning, Optional<Parti
                     false);
         }
 
+        public static Global partitionedOnExpression(Partitioning nodePartitioning, Collection<JoinNode.EquiJoinClause> equalities)
```

Comment:
The assembly of CoalesceExpressions should be done in the caller (PropertyDerivations.visitJoin). That's the place that's responsible for determining the properties guaranteed by each operator.

Also, ActualProperties should not have a dependency on EquiJoinClauses.